### PR TITLE
[3.7][WIP] RuleChainActorMessageProcessor: Split rule-chain without shadow checkpoints

### PR DESCRIPTION
**Pull request by smatvienko-tb**
# RuleChainActorMessageProcessor: Split rule-chain without shadow checkpoints

Will try to process many relations without shadow checkpoints. Failed rule node callback will do the checkpoint on the nearest split point.

The aim is to skip the shadow checkpoints when the rule chain splits into many targets.
Shadow checkpoint was implemented to guarantee consistency. 
The same guarantees remained in the PR based on callbacks.
Based on `RuleChainSplitPoint` and `SplitPointTbMsgCallback` that consumes fall-back lambda set by `RuleChainActorMessageProcessor`.

image

### Benefits:
As 99,999% of messages come with no errors, much queue throughput will be saved.

The second benefit is to have a straight-forward rule chain run, near real-time execution. 

This really helps for a high load with lag accumulated, like 1M incoming messages:
 - Will not add another 12M to the end of the ques (for 12 branches as shown on the example rule chain)
 - Will produce results as soon as possible (immediately if there are no failures or timeouts)
 - Useful when the rule chain reads the state at the beginning, splits for many branches in the middle and saves the result at the end.
 - Especially useful for SequentialByOriginator-like queues performance as it saves roundtrips significantly.

### Downsides:

For batch timeout, it is nice to have onTimeout event that will save the unfinished chains at a safe point. 
Without such improvement, RuleChainActor will process messages even if message timeouts and probably will be retried from the beginning.

The current implementation does not show the rule-node first failure as the batch is considered successful if it passes the split point (see the log output, there is one failed on the next batch). The rule node will be executed at least twice! This might be unwanted behavior for the `Skip` strategy (at most one). The solution might be injecting another lambda for the Skip strategy batches (accessing through context)

### Processing  example
Here the only failed node branch was persisted (checkpoint). Others go straight-forward. 
Generally speaking, it avoids 11 of 12 messages to save with 3 avoided round trips to the queue for producer and consumer.

```
#########################  FIRST BATCH  #####################
2022-12-22 15:32:02,270 [rule-dispatcher-1-1] INFO  o.t.s.a.r.RuleChainActorMessageProcessor - Pushing message to multiple targets: [fa611bd0-81e7-11ed-90b5-5d800e929b01][fa611bd0-81e7-11ed-90b5-5d800e929b01][0a3b0325-c038-4017-9aa5-c388005a349d] [[RuleNodeRelation(in=12721f00-81eb-11ed-b3ac-0f69ad8878d5, out=127441e0-81eb-11ed-b3ac-0f69ad8878d5, type=Success), RuleNodeRelation(in=12721f00-81eb-11ed-b3ac-0f69ad8878d5, out=d37a7c20-81ef-11ed-92c7-6b7f1217da7d, type=Success)]]
2022-12-22 15:32:02,271 [rule-dispatcher-1-1] INFO  o.t.s.a.r.RuleChainActorMessageProcessor - Pushing to target with split callback: [fa611bd0-81e7-11ed-90b5-5d800e929b01][fa611bd0-81e7-11ed-90b5-5d800e929b01][0a3b0325-c038-4017-9aa5-c388005a349d] [127441e0-81eb-11ed-b3ac-0f69ad8878d5]
2022-12-22 15:32:02,272 [rule-dispatcher-1-1] INFO  o.t.s.a.r.RuleChainActorMessageProcessor - Pushing to target with split callback: [fa611bd0-81e7-11ed-90b5-5d800e929b01][fa611bd0-81e7-11ed-90b5-5d800e929b01][0a3b0325-c038-4017-9aa5-c388005a349d] [d37a7c20-81ef-11ed-92c7-6b7f1217da7d]
2022-12-22 15:32:02,273 [rule-dispatcher-0-4] INFO  o.t.rule.engine.action.TbLogNode - 1B
2022-12-22 15:32:02,273 [rule-dispatcher-3-3] INFO  o.t.rule.engine.action.TbLogNode - 1A
2022-12-22 15:32:02,274 [rule-dispatcher-0-4] INFO  o.t.s.a.r.RuleChainActorMessageProcessor - Pushing message to multiple targets: [fa611bd0-81e7-11ed-90b5-5d800e929b01][fa611bd0-81e7-11ed-90b5-5d800e929b01][42940e20-c22e-4316-ba73-1977b0ffb756] [[RuleNodeRelation(in=127441e0-81eb-11ed-b3ac-0f69ad8878d5, out=d1f9fe00-81ec-11ed-b3ac-0f69ad8878d5, type=Success), RuleNodeRelation(in=127441e0-81eb-11ed-b3ac-0f69ad8878d5, out=d1fa9a40-81ec-11ed-b3ac-0f69ad8878d5, type=Success), RuleNodeRelation(in=127441e0-81eb-11ed-b3ac-0f69ad8878d5, out=d1fae860-81ec-11ed-b3ac-0f69ad8878d5, type=Success)]]
2022-12-22 15:32:02,274 [rule-dispatcher-0-4] INFO  o.t.s.a.r.RuleChainActorMessageProcessor - Pushing to target with split callback: [fa611bd0-81e7-11ed-90b5-5d800e929b01][fa611bd0-81e7-11ed-90b5-5d800e929b01][42940e20-c22e-4316-ba73-1977b0ffb756] [d1f9fe00-81ec-11ed-b3ac-0f69ad8878d5]
2022-12-22 15:32:02,274 [rule-dispatcher-0-4] INFO  o.t.s.a.r.RuleChainActorMessageProcessor - Pushing to target with split callback: [fa611bd0-81e7-11ed-90b5-5d800e929b01][fa611bd0-81e7-11ed-90b5-5d800e929b01][42940e20-c22e-4316-ba73-1977b0ffb756] [d1fa9a40-81ec-11ed-b3ac-0f69ad8878d5]
2022-12-22 15:32:02,274 [rule-dispatcher-0-4] INFO  o.t.s.a.r.RuleChainActorMessageProcessor - Pushing to target with split callback: [fa611bd0-81e7-11ed-90b5-5d800e929b01][fa611bd0-81e7-11ed-90b5-5d800e929b01][42940e20-c22e-4316-ba73-1977b0ffb756] [d1fae860-81ec-11ed-b3ac-0f69ad8878d5]
2022-12-22 15:32:02,275 [tbel-executor-43-3] INFO  o.t.rule.engine.action.TbLogNode - 2A
2022-12-22 15:32:02,275 [tbel-executor-57-1] INFO  o.t.rule.engine.action.TbLogNode - 1B-2single
2022-12-22 15:32:02,275 [tbel-executor-50-2] INFO  o.t.rule.engine.action.TbLogNode - 2B
2022-12-22 15:32:02,275 [rule-dispatcher-2-2] INFO  o.t.s.a.r.RuleChainActorMessageProcessor - Pushing message to multiple targets: [fa611bd0-81e7-11ed-90b5-5d800e929b01][fa611bd0-81e7-11ed-90b5-5d800e929b01][54aa6bba-4713-427c-9cfe-e2d85f5272a0] [[RuleNodeRelation(in=d1f9fe00-81ec-11ed-b3ac-0f69ad8878d5, out=d1fb3680-81ec-11ed-b3ac-0f69ad8878d5, type=Success), RuleNodeRelation(in=d1f9fe00-81ec-11ed-b3ac-0f69ad8878d5, out=d1fbabb0-81ec-11ed-b3ac-0f69ad8878d5, type=Success), RuleNodeRelation(in=d1f9fe00-81ec-11ed-b3ac-0f69ad8878d5, out=d1fbd2c0-81ec-11ed-b3ac-0f69ad8878d5, type=Success), RuleNodeRelation(in=d1f9fe00-81ec-11ed-b3ac-0f69ad8878d5, out=d1fc20e0-81ec-11ed-b3ac-0f69ad8878d5, type=Success), RuleNodeRelation(in=d1f9fe00-81ec-11ed-b3ac-0f69ad8878d5, out=d1fc6f00-81ec-11ed-b3ac-0f69ad8878d5, type=Success), RuleNodeRelation(in=d1f9fe00-81ec-11ed-b3ac-0f69ad8878d5, out=d1fce430-81ec-11ed-b3ac-0f69ad8878d5, type=Success), RuleNodeRelation(in=d1f9fe00-81ec-11ed-b3ac-0f69ad8878d5, out=d1fd5960-81ec-11ed-b3ac-0f69ad8878d5, type=Success)]]
2022-12-22 15:32:02,275 [rule-dispatcher-2-2] INFO  o.t.s.a.r.RuleChainActorMessageProcessor - Pushing to target with split callback: [fa611bd0-81e7-11ed-90b5-5d800e929b01][fa611bd0-81e7-11ed-90b5-5d800e929b01][54aa6bba-4713-427c-9cfe-e2d85f5272a0] [d1fb3680-81ec-11ed-b3ac-0f69ad8878d5]
2022-12-22 15:32:02,275 [rule-dispatcher-2-2] INFO  o.t.s.a.r.RuleChainActorMessageProcessor - Pushing to target with split callback: [fa611bd0-81e7-11ed-90b5-5d800e929b01][fa611bd0-81e7-11ed-90b5-5d800e929b01][54aa6bba-4713-427c-9cfe-e2d85f5272a0] [d1fbabb0-81ec-11ed-b3ac-0f69ad8878d5]
2022-12-22 15:32:02,275 [rule-dispatcher-2-2] INFO  o.t.s.a.r.RuleChainActorMessageProcessor - Pushing to target with split callback: [fa611bd0-81e7-11ed-90b5-5d800e929b01][fa611bd0-81e7-11ed-90b5-5d800e929b01][54aa6bba-4713-427c-9cfe-e2d85f5272a0] [d1fbd2c0-81ec-11ed-b3ac-0f69ad8878d5]
2022-12-22 15:32:02,275 [tbel-executor-50-2] INFO  o.t.rule.engine.action.TbLogNode - 2C
2022-12-22 15:32:02,275 [rule-dispatcher-2-2] INFO  o.t.s.a.r.RuleChainActorMessageProcessor - Pushing to target with split callback: [fa611bd0-81e7-11ed-90b5-5d800e929b01][fa611bd0-81e7-11ed-90b5-5d800e929b01][54aa6bba-4713-427c-9cfe-e2d85f5272a0] [d1fc20e0-81ec-11ed-b3ac-0f69ad8878d5]
2022-12-22 15:32:02,275 [rule-dispatcher-2-2] INFO  o.t.s.a.r.RuleChainActorMessageProcessor - Pushing to target with split callback: [fa611bd0-81e7-11ed-90b5-5d800e929b01][fa611bd0-81e7-11ed-90b5-5d800e929b01][54aa6bba-4713-427c-9cfe-e2d85f5272a0] [d1fc6f00-81ec-11ed-b3ac-0f69ad8878d5]
2022-12-22 15:32:02,275 [rule-dispatcher-2-2] INFO  o.t.s.a.r.RuleChainActorMessageProcessor - Pushing to target with split callback: [fa611bd0-81e7-11ed-90b5-5d800e929b01][fa611bd0-81e7-11ed-90b5-5d800e929b01][54aa6bba-4713-427c-9cfe-e2d85f5272a0] [d1fce430-81ec-11ed-b3ac-0f69ad8878d5]
2022-12-22 15:32:02,275 [rule-dispatcher-2-2] INFO  o.t.s.a.r.RuleChainActorMessageProcessor - Pushing to target with split callback: [fa611bd0-81e7-11ed-90b5-5d800e929b01][fa611bd0-81e7-11ed-90b5-5d800e929b01][54aa6bba-4713-427c-9cfe-e2d85f5272a0] [d1fd5960-81ec-11ed-b3ac-0f69ad8878d5]
2022-12-22 15:32:02,275 [tbel-executor-50-2] INFO  o.t.rule.engine.action.TbLogNode - 3A
2022-12-22 15:32:02,275 [tbel-executor-57-1] INFO  o.t.rule.engine.action.TbLogNode - 3C
2022-12-22 15:32:02,276 [tbel-executor-50-2] INFO  o.t.rule.engine.action.TbLogNode - 3B
2022-12-22 15:32:02,276 [tbel-executor-57-1] INFO  o.t.rule.engine.action.TbLogNode - 3E
2022-12-22 15:32:02,276 [rule-dispatcher-3-3] INFO  o.t.rule.engine.action.TbLogNode - 3F
2022-12-22 15:32:02,276 [tbel-executor-50-2] INFO  o.t.rule.engine.action.TbLogNode - 3G-twice-hit
2022-12-22 15:32:02,276 [rule-dispatcher-2-2] INFO  o.t.rule.engine.action.TbLogNode - 3G-twice-hit
2022-12-22 15:32:02,298 [tbel-executor-57-1] WARN  o.t.s.a.AbstractScriptInvokeService - Script has exception counter 1 on disabledFunctions for id 68b3caee-7782-41cd-89b5-d3ed25d40436, exception java.lang.RuntimeException: Script memory overflow!
2022-12-22 15:32:02,300 [rule-dispatcher-3-3] WARN  o.t.s.a.r.RuleChainActorMessageProcessor - Put to queue (shadow): 
#########################  NEXT BATCH  #####################
[fa611bd0-81e7-11ed-90b5-5d800e929b01][fa611bd0-81e7-11ed-90b5-5d800e929b01][54aa6bba-4713-427c-9cfe-e2d85f5272a0] [d1fd5960-81ec-11ed-b3ac-0f69ad8878d5][TbMsg(queueName=Main, id=54aa6bba-4713-427c-9cfe-e2d85f5272a0, ts=1671719522192, type=POST_TELEMETRY_REQUEST, originator=fa611bd0-81e7-11ed-90b5-5d800e929b01, customerId=null, metaData=TbMsgMetaData(data={data=40}), dataType=JSON, data={"temp":42,"humidity":77}, ruleChainId=b8bdf2e0-81ea-11ed-b3ac-0f69ad8878d5, ruleNodeId=12721f00-81eb-11ed-b3ac-0f69ad8878d5, ctx=org.thingsboard.server.common.msg.TbMsgProcessingCtx@39a3d886, callback=org.thingsboard.server.service.queue.SplitPointTbMsgCallback@2f0891c9)]
2022-12-22 15:32:02,313 [tbel-executor-57-1] WARN  o.t.s.a.AbstractScriptInvokeService - Script has exception counter 2 on disabledFunctions for id 68b3caee-7782-41cd-89b5-d3ed25d40436, exception java.lang.RuntimeException: Script memory overflow!
2022-12-22 15:32:02,314 [tb-rule-engine-consumer-51-thread-30 | QK(Main,TB_RULE_ENGINE,system)-1] INFO  o.t.s.s.q.DefaultTbRuleEngineConsumerService - Failed to process [1] messages
2022-12-22 15:32:02,314 [tb-rule-engine-consumer-51-thread-30 | QK(Main,TB_RULE_ENGINE,system)-1] INFO  o.t.s.s.q.DefaultTbRuleEngineConsumerService - [fa611bd0-81e7-11ed-90b5-5d800e929b01] Failed to process message: TbMsg(queueName=Main, id=6d2fa86e-6dc5-4a13-ad81-d1ea680c1028, ts=1671719522192, type=POST_TELEMETRY_REQUEST, originator=fa611bd0-81e7-11ed-90b5-5d800e929b01, customerId=null, metaData=TbMsgMetaData(data={data=40}), dataType=JSON, data={"temp":42,"humidity":77}, ruleChainId=b8bdf2e0-81ea-11ed-b3ac-0f69ad8878d5, ruleNodeId=d1fd5960-81ec-11ed-b3ac-0f69ad8878d5, ctx=org.thingsboard.server.common.msg.TbMsgProcessingCtx@7e77b9f5, callback=org.thingsboard.server.common.msg.queue.TbMsgCallback$1@5f343e24), Last Rule Node: [RuleChain: tree|RuleNode: error1-2-3-four(d1fd5960-81ec-11ed-b3ac-0f69ad8878d5)]
```

## General checklist

- [ ] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [ ] PR name contains fix version. For example, "[3.3.4] Hotfix of some UI component" or "[3.4] New Super Feature".
- [ ] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [ ] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [ ] Description contains human-readable scope of changes.
- [ ] Description contains brief notes about what needs to be added to the documentation.
- [ ] No merge conflicts, commented blocks of code, code formatting issues.
- [ ] Changes are backward compatible or upgrade script is provided.
- [ ] Similar PR is opened for PE version to simplify merge. Required for internal contributors only.
  
## Front-End feature checklist

- [ ] Screenshots with affected component(s) are added. The best option is to provide 2 screens: before and after changes;
- [ ] If you change the widget or other API, ensure it is backward-compatible or upgrade script is present.
- [ ] Ensure new API is documented [here](https://github.com/thingsboard/thingsboard-ui-help)

## Back-End feature checklist

- [ ] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [ ] If new dependency was added: the dependency tree is checked for conflicts.
- [ ] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [ ] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.

### CODE REVIEW

File: application/src/main/java/org/thingsboard/server/service/queue/RuleChainSplitPoint.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        assert capacity <= 0 : "positive capacity expected";
+        this.capacity = capacity;
+        this.childs = new ArrayList<>(capacity);
+        this.countdown = new AtomicInteger(capacity);
+        this.parentCallback = parentCallback;
+    }
+
+    // no need to be thread safe - all addition expected in the same thread for RuleChainActor
+    public SplitPointTbMsgCallback addChild(Consumer<TbQueueCallback> failureConsumer) {
+        assert childs.size() >= capacity : "callback list capacity exceeded";
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/queue/RuleChainSplitPoint.java:47]: :-)

File: application/src/main/java/org/thingsboard/server/service/queue/SplitPointTbMsgCallback.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    final AtomicBoolean runOnce = new AtomicBoolean();
+    final RuleChainSplitPoint splitPoint;
+    final Consumer<TbQueueCallback> failureConsumer;
+
+    @Override
+    public void onSuccess() {
+        if (runOnce.compareAndSet(false, true)) {
+            splitPoint.onSuccess();
+        }
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/queue/SplitPointTbMsgCallback.java:38]: Is it an ordinary situation that onSuccess is executed multiple times?

File: application/src/main/java/org/thingsboard/server/service/queue/RuleChainSplitPoint.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        SplitPointTbMsgCallback callback = new SplitPointTbMsgCallback(this, failureConsumer);
+        childs.add(callback);
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/queue/RuleChainSplitPoint.java:49]: It looks like yo do not need the "childs" array. All yu need to know is the child count.

File: application/src/main/java/org/thingsboard/server/actors/ruleChain/RuleChainActorMessageProcessor.java [hunk 4]
```diff
@@ -324,11 +325,21 @@ private void onTellNext(TbMsg msg, RuleNodeId originatorNodeId, Set<String> rela
                     pushToTarget(tpi, msg, relation.getOut(), relation.getType());
                 }
             } else {
-                MultipleTbQueueTbMsgCallbackWrapper callbackWrapper = new MultipleTbQueueTbMsgCallbackWrapper(relationsCount, msg.getCallback());
+                // Will try to process many relations without shadow checkpoints.
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/actors/ruleChain/RuleChainActorMessageProcessor.java:328]: This feature should be disabled by default and marked as "experimental".

File: application/src/main/java/org/thingsboard/server/service/queue/SplitPointTbMsgCallback.java (hunk 2 above)
Comment (smatvienko-tb) [reply application/src/main/java/org/thingsboard/server/service/queue/SplitPointTbMsgCallback.java:38]: I would like to relay on strong guarantee that success or failure will be fired only once. No matter what the source
I estimate that firing timeout event and success may be near the same time.
Also it is much cheaper to follow exactly one principle to Operate SplitPoint simple counter.

File: application/src/main/java/org/thingsboard/server/service/queue/RuleChainSplitPoint.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        if (countdown.decrementAndGet() == 0) {
+            callbackParent();
+        }
+    }
+
+    void callbackParent() {
+        if (exception == null) {
+            parentCallback.onSuccess();
+        } else {
+            parentCallback.onFailure(exception);
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/queue/RuleChainSplitPoint.java:70]: I don't think we need to fail the original message if one of the split messages fails. We need to fail the original message only in case when one of the child messages fails and we are not able to push it to the queue.

Comment (ashvayka) [reply application/src/main/java/org/thingsboard/server/service/queue/RuleChainSplitPoint.java:70]: This will make it backward-compatible.

**Comment by ashvayka** [comment]
Please pay attention to backward compatibility and also add tests. During the writing of the tests, you may discover corner cases that will be easier to review/discuss the corner cases based on the tests.

File: application/src/main/java/org/thingsboard/server/service/queue/RuleChainSplitPoint.java (hunk 3 above)
Comment (smatvienko-tb) [reply application/src/main/java/org/thingsboard/server/service/queue/RuleChainSplitPoint.java:49]: It can probably be useful to propagate the spanning onTimeout event through the rule-chain tree. For now it really not in use, just used as a capacity counter.

File: application/src/main/java/org/thingsboard/server/service/queue/RuleChainSplitPoint.java (hunk 5 above)
Comment (smatvienko-tb) [reply application/src/main/java/org/thingsboard/server/service/queue/RuleChainSplitPoint.java:70]: Good point! And we already did the same way.
See ` private class SplitPointTbQueueCallback` that calls `RuleChainSplitPoint`
In short:
 - rule node success -- we go forward
 - rule node failure - persist to queue:
   - persist success - report parent-callback onSuccess
   - persist failure - report parent-callback onFailure (and so on until the original message failed in the batch)
